File: scripts/generate-pricing-sheet.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Iterable

from PIL import Image, ImageDraw, ImageFont


ROOT_DIR = Path(__file__).resolve().parent.parent
DATA_PATH = ROOT_DIR / "src" / "data" / "services.content.json"
OUTPUT_DIR = ROOT_DIR / "public" / "downloads"
PNG_PATH = OUTPUT_DIR / "glowwithvani-pricing.png"
PDF_PATH = OUTPUT_DIR / "glowwithvani-pricing.pdf"

PALETTE = {
    "background": "#faf7f2",
    "surface": "#fffdf9",
    "surface_soft": "#f5efe5",
    "ink": "#241b16",
    "muted": "#6c5f51",
    "accent": "#8c6a42",
    "line": "#ddcfc0",
    "glow": "#efe3d1",
}

WIDTH = 1600
OUTER_PAD = 96
CARD_GAP = 28
COLS = 2
CARD_WIDTH = (WIDTH - OUTER_PAD * 2 - CARD_GAP) // COLS
CARD_HEIGHT = 360
HEADER_HEIGHT = 220
NOTES_HEIGHT = 220


def load_font(size: int, *, serif: bool = False, bold: bool = False) -> ImageFont.FreeTypeFont | ImageFont.ImageFont:
    candidates: list[str]
    if serif:
        candidates = [
            "/System/Library/Fonts/Supplemental/Georgia Bold.ttf" if bold else "/System/Library/Fonts/Supplemental/Georgia.ttf",
            "/System/Library/Fonts/Supplemental/Times New Roman Bold.ttf" if bold else "/System/Library/Fonts/Supplemental/Times New Roman.ttf",
            "/usr/share/fonts/truetype/dejavu/DejaVuSerif-Bold.ttf" if bold else "/usr/share/fonts/truetype/dejavu/DejaVuSerif.ttf",
            "/usr/share/fonts/truetype/liberation2/LiberationSerif-Bold.ttf" if bold else "/usr/share/fonts/truetype/liberation2/LiberationSerif-Regular.ttf",
            "/usr/share/fonts/liberation/LiberationSerif-Bold.ttf" if bold else "/usr/share/fonts/liberation/LiberationSerif-Regular.ttf",
        ]
    else:
        candidates = [
            "/System/Library/Fonts/Supplemental/Arial Bold.ttf" if bold else "/System/Library/Fonts/Supplemental/Arial.ttf",
            "/System/Library/Fonts/Supplemental/Helvetica.ttc",
            "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf" if bold else "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf",
            "/usr/share/fonts/truetype/liberation2/LiberationSans-Bold.ttf" if bold else "/usr/share/fonts/truetype/liberation2/LiberationSans-Regular.ttf",
            "/usr/share/fonts/liberation/LiberationSans-Bold.ttf" if bold else "/usr/share/fonts/liberation/LiberationSans-Regular.ttf",
        ]

    candidates.extend(
        [
            "/System/Library/Fonts/Supplemental/Arial.ttf",
            "/System/Library/Fonts/Supplemental/Times New Roman.ttf",
        ]
    )

    for candidate in candidates:
        try:
            return ImageFont.truetype(candidate, size=size)
        except OSError:
            continue

    raise RuntimeError(f"No scalable font found for {'serif' if serif else 'sans'} {'bold' if bold else 'regular'} text")


FONT_KICKER = load_font(28, bold=True)
FONT_TITLE = load_font(76, serif=True)
FONT_SUBTITLE = load_font(28)
FONT_CARD_TITLE = load_font(38, serif=True)
FONT_CARD_PRICE = load_font(30, bold=True)
FONT_CARD_BODY = load_font(23)
FONT_NOTES_HEADING = load_font(24, bold=True)
FONT_NOTES = load_font(22)


def text_size(draw: ImageDraw.ImageDraw, text: str, font: ImageFont.ImageFont) -> tuple[int, int]:
    bbox = draw.textbbox((0, 0), text, font=font)
    return bbox[2] - bbox[0], bbox[3] - bbox[1]


def wrap_text(draw: ImageDraw.ImageDraw, text: str, font: ImageFont.ImageFont, max_width: int) -> list[str]:
    words = text.split()
    lines: list[str] = []
    current = ""

    for word in words:
        candidate = f"{current} {word}".strip()
        if text_size(draw, candidate, font)[0] <= max_width:
            current = candidate
        else:
            if current:
                lines.append(current)
            current = word

    if current:
        lines.append(current)

    return lines


def draw_wrapped_text(
    draw: ImageDraw.ImageDraw,
    *,
    x: int,
    y: int,
    lines: Iterable[str],
    font: ImageFont.ImageFont,
    fill: str,
    line_height: int,
) -> int:
    current_y = y
    for line in lines:
        draw.text((x, current_y), line, font=font, fill=fill)
        current_y += line_height
    return current_y


with DATA_PATH.open() as handle:
    content = json.load(handle)

services = content["services"]
service_pricing_notes = content["servicePricingNotes"]

row_count = (len(services) + COLS - 1) // COLS
grid_height = row_count * CARD_HEIGHT + max(0, row_count - 1) * CARD_GAP
HEIGHT = OUTER_PAD * 2 + HEADER_HEIGHT + grid_height + NOTES_HEIGHT

image = Image.new("RGB", (WIDTH, HEIGHT), PALETTE["background"])
draw = ImageDraw.Draw(image)

# Soft background accents.
draw.ellipse((WIDTH - 440, 10, WIDTH - 80, 370), fill=PALETTE["glow"])
draw.ellipse((WIDTH - 700, HEIGHT - 620, WIDTH - 260, HEIGHT - 180), fill="#f2e6d8")

# Header.
draw.text((OUTER_PAD, 92), "GLOWWITHVANI", font=FONT_KICKER, fill=PALETTE["accent"])
draw.text((OUTER_PAD, 140), "Services & Pricing", font=FONT_TITLE, fill=PALETTE["ink"])
draw.text(
    (OUTER_PAD, 228),
    "Luxury makeup services for bridal, event, and editorial beauty.",
    font=FONT_SUBTITLE,
    fill=PALETTE["muted"],
)

for index, service in enumerate(services):
    col = index % COLS
    row = index // COLS
    x = OUTER_PAD + col * (CARD_WIDTH + CARD_GAP)
    y = OUTER_PAD + HEADER_HEIGHT + row * (CARD_HEIGHT + CARD_GAP)

    draw.rounded_rectangle(
        (x, y, x + CARD_WIDTH, y + CARD_HEIGHT),
        radius=28,
        fill=PALETTE["surface"],
        outline=PALETTE["line"],
        width=2,
    )

    inner_x = x + 40
    inner_y = y + 36
    max_text_width = CARD_WIDTH - 80

    title_lines = wrap_text(draw, service["title"], FONT_CARD_TITLE, max_text_width)
    after_title_y = draw_wrapped_text(
        draw,
        x=inner_x,
        y=inner_y,
        lines=title_lines,
        font=FONT_CARD_TITLE,
        fill=PALETTE["ink"],
        line_height=44,
    )

    price_y = after_title_y + 18
    draw.text((inner_x, price_y), service["startingPrice"], font=FONT_CARD_PRICE, fill=PALETTE["accent"])

    _, price_height = text_size(draw, service["startingPrice"], FONT_CARD_PRICE)
    description_y = price_y + price_height + 26
    description_lines = wrap_text(draw, service["description"], FONT_CARD_BODY, max_text_width)[:4]
    draw_wrapped_text(
        draw,
        x=inner_x,
        y=description_y,
        lines=description_lines,
        font=FONT_CARD_BODY,
        fill=PALETTE["muted"],
        line_height=31,
    )

notes_y = OUTER_PAD + HEADER_HEIGHT + grid_height + 36
draw.rounded_rectangle(
    (OUTER_PAD, notes_y, WIDTH - OUTER_PAD, notes_y + 164),
    radius=28,
    fill=PALETTE["surface_soft"],
    outline=PALETTE["line"],
    width=2,
)
draw.text((OUTER_PAD + 28, notes_y + 24), "NOTES", font=FONT_NOTES_HEADING, fill=PALETTE["accent"])

for index, note in enumerate(service_pricing_notes):
    line_y = notes_y + 70 + index * 36
    draw.ellipse((OUTER_PAD + 28, line_y + 8, OUTER_PAD + 36, line_y + 16), fill=PALETTE["accent"])
    draw.text((OUTER_PAD + 52, line_y), note, font=FONT_NOTES, fill=PALETTE["muted"])

draw.text(
    (OUTER_PAD + 28, notes_y + 132),
    "Final pricing may vary based on look, location, and custom requirements.",
    font=FONT_NOTES,
    fill=PALETTE["muted"],
)

OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
image.save(PNG_PATH, format="PNG")
image.save(PDF_PATH, format="PDF", resolution=300)

print(f"Generated {PNG_PATH}")
print(f"Generated {PDF_PATH}")

File: scripts/generate-pricing-sheet.mjs
import fs from "node:fs";
import path from "node:path";
import sharp from "sharp";
import PDFDocument from "pdfkit";

const rootDir = process.cwd();
const dataPath = path.join(rootDir, "src", "data", "services.content.json");
const outputDir = path.join(rootDir, "public", "downloads");
const pngPath = path.join(outputDir, "glowwithvani-pricing.png");
const pdfPath = path.join(outputDir, "glowwithvani-pricing.pdf");

const palette = {
  background: "#faf7f2",
  surface: "#fffdf9",
  surfaceSoft: "#f5efe5",
  ink: "#241b16",
  muted: "#6c5f51",
  accent: "#8c6a42",
  line: "#ddcfc0",
  glow: "#efe3d1"
};

const WIDTH = 1600;
const OUTER_PAD = 96;
const CARD_GAP = 28;
const COLS = 2;
const CARD_WIDTH = (WIDTH - OUTER_PAD * 2 - CARD_GAP) / COLS;
const CARD_HEIGHT = 360;
const HEADER_HEIGHT = 220;
const NOTES_HEIGHT = 220;

const titleCharFactor = 18.5;
const bodyCharFactor = 11.5;

function escapeXml(value) {
  return String(value)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;")
    .replaceAll('"', "&quot;")
    .replaceAll("'", "&apos;");
}

function wrapText(text, maxWidth, charFactor) {
  const maxChars = Math.max(10, Math.floor(maxWidth / charFactor));
  const words = text.split(/\s+/);
  const lines = [];
  let current = "";

  for (const word of words) {
    const candidate = current ? `${current} ${word}` : word;
    if (candidate.length <= maxChars) {
      current = candidate;
    } else {
      if (current) {
        lines.push(current);
      }
      current = word;
    }
  }

  if (current) {
    lines.push(current);
  }

  return lines;
}

function textBlock(lines, x, y, fontSize, fill, lineHeight, extra = "") {
  return lines
    .map(
      (line, index) =>
        `<text x="${x}" y="${y + index * lineHeight}" font-size="${fontSize}" fill="${fill}" ${extra}>${escapeXml(line)}</text>`
    )
    .join("");
}

async function main() {
  const content = JSON.parse(fs.readFileSync(dataPath, "utf8"));
  const services = content.services;
  const servicePricingNotes = content.servicePricingNotes;

  const rowCount = Math.ceil(services.length / COLS);
  const gridHeight = rowCount * CARD_HEIGHT + Math.max(0, rowCount - 1) * CARD_GAP;
  const height = OUTER_PAD * 2 + HEADER_HEIGHT + gridHeight + NOTES_HEIGHT;

  const cardInnerWidth = CARD_WIDTH - 80;

  let cardsSvg = "";
  for (const [index, service] of services.entries()) {
    const col = index % COLS;
    const row = Math.floor(index / COLS);
    const x = OUTER_PAD + col * (CARD_WIDTH + CARD_GAP);
    const y = OUTER_PAD + HEADER_HEIGHT + row * (CARD_HEIGHT + CARD_GAP);
    const innerX = x + 40;
    const innerY = y + 52;

    const titleLines = wrapText(service.title, cardInnerWidth, titleCharFactor);
    const titleLineHeight = 46;
    const titleBottomY = innerY + (titleLines.length - 1) * titleLineHeight;
    const priceY = titleBottomY + 58;
    const bodyLines = wrapText(service.description, cardInnerWidth, bodyCharFactor).slice(0, 4);
    const bodyY = priceY + 42;

    cardsSvg += `
      <rect x="${x}" y="${y}" rx="28" ry="28" width="${CARD_WIDTH}" height="${CARD_HEIGHT}" fill="${palette.surface}" stroke="${palette.line}" stroke-width="2" />
      ${textBlock(titleLines, innerX, innerY, 38, palette.ink, titleLineHeight, 'font-family="Georgia, Times New Roman, serif"')}
      <text x="${innerX}" y="${priceY}" font-size="30" fill="${palette.accent}" font-family="Arial, Helvetica, sans-serif" font-weight="700">${escapeXml(service.startingPrice)}</text>
      ${textBlock(bodyLines, innerX, bodyY, 23, palette.muted, 32, 'font-family="Arial, Helvetica, sans-serif"')}
    `;
  }

  const notesY = OUTER_PAD + HEADER_HEIGHT + gridHeight + 36;
  const notesLines = servicePricingNotes
    .map((note, index) => {
      const lineY = notesY + 78 + index * 38;
      return `
        <circle cx="${OUTER_PAD + 32}" cy="${lineY - 7}" r="5" fill="${palette.accent}" />
        <text x="${OUTER_PAD + 52}" y="${lineY}" font-size="22" fill="${palette.muted}" font-family="Arial, Helvetica, sans-serif">${escapeXml(note)}</text>
      `;
    })
    .join("");

  const svg = `
    <svg width="${WIDTH}" height="${height}" viewBox="0 0 ${WIDTH} ${height}" xmlns="http://www.w3.org/2000/svg">
      <rect width="${WIDTH}" height="${height}" fill="${palette.background}" />
      <circle cx="${WIDTH - 260}" cy="190" r="180" fill="${palette.glow}" />
      <circle cx="${WIDTH - 480}" cy="${height - 400}" r="220" fill="#f2e6d8" />

      <text x="${OUTER_PAD}" y="120" font-size="28" fill="${palette.accent}" font-family="Arial, Helvetica, sans-serif" font-weight="700" letter-spacing="1">GLOWWITHVANI</text>
      <text x="${OUTER_PAD}" y="200" font-size="76" fill="${palette.ink}" font-family="Georgia, Times New Roman, serif">Services &amp; Pricing</text>
      <text x="${OUTER_PAD}" y="252" font-size="28" fill="${palette.muted}" font-family="Arial, Helvetica, sans-serif">Luxury makeup services for bridal, event, and editorial beauty.</text>

      ${cardsSvg}

      <rect x="${OUTER_PAD}" y="${notesY}" rx="28" ry="28" width="${WIDTH - OUTER_PAD * 2}" height="164" fill="${palette.surfaceSoft}" stroke="${palette.line}" stroke-width="2" />
      <text x="${OUTER_PAD + 28}" y="${notesY + 38}" font-size="24" fill="${palette.accent}" font-family="Arial, Helvetica, sans-serif" font-weight="700">NOTES</text>
      ${notesLines}
      <text x="${OUTER_PAD + 28}" y="${notesY + 142}" font-size="22" fill="${palette.muted}" font-family="Arial, Helvetica, sans-serif">Final pricing may vary based on look, location, and custom requirements.</text>
    </svg>
  `;

  fs.mkdirSync(outputDir, { recursive: true });
  const pngBuffer = await sharp(Buffer.from(svg)).png().toBuffer();
  fs.writeFileSync(pngPath, pngBuffer);

  await new Promise((resolve) => {
    const doc = new PDFDocument({ size: [WIDTH, height], margin: 0 });
    const stream = fs.createWriteStream(pdfPath);
    doc.pipe(stream);
    doc.image(pngBuffer, 0, 0, { width: WIDTH, height });
    doc.end();
    stream.on("finish", resolve);
  });

  console.log(`Generated ${pngPath}`);
  console.log(`Generated ${pdfPath}`);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
